back/app/services/xlsx_service.rb

```diff
@@ -267,15 +267,22 @@ def custom_field_columns(record_to_user, view_private_attributes)
     user_custom_fields&.map do |field|
 
       column_name = multiloc_service.t(field.title_multiloc)
-      value_getter = # lambda that gets a record and returns the field value
+      value_getter = #lambda that gets a record and returns the field value
         if field.key == 'domicile'  # 'domicile' is a special case
           lambda do |record|
             user = record.send(record_to_user)
             multiloc_service.t(areas[user.domicile]&.title_multiloc) if user && user.custom_field_values['domicile']
           end
+        elsif field.support_options? #field with option
+          lambda do |record|
+            user = record.send(record_to_user)
+
+            user && user.custom_field_values[field.key] &&  multiloc_service.t(field.custom_field_options.where(key: user.custom_field_values[field.key]).first.title_multiloc)
```

Comment:
This code is in a critical performance path, in the sense that we can (and often are) exporting files with 1000s of records. The lambda is executed for every row.

This causes an N+1 query, triggering a query for every row. Now, it's probably not _so_ bad, because ActiveRecord will cache identical queries, but it there are many custom_field_options (we have some platforms with 1000s as well) it will still be very bad. This is not a theoretical issue, we've had N+1's being introduced in the xlsx exports before, breaking the functionality for larger platforms.

To avoid this, we should do the same as we're doing with areas in the same method: Do a single query for all custom_field_options upfront, and then in Ruby, and not SQL, get the value out.

Reply:
You're right, thanks for the feedback ! I now followed the same approach as areas, except the areas are referenced by id and custom fields by key (which I just don't understand why, was it for this use-case of exposing it to the admin ? Now that's removed, could we move back to ids?)

Reply:
Yes, the key was for the use case of controlling what's in the actual data. If we completely hide this key field, it's indeed counter intuitive, but I still think it will make sense for future scenarios where more data control is needed (e.g. API integrations) so I would keep it. Also saves a tricky data migration.